rfcs/20210129-optimize-pd-creation/README.md

```diff
@@ -0,0 +1,262 @@
+    std::type_index impl_id_;
+    std::vector<memory_desc_t> bwd_mds;
+
+    // Obtained from `op_desc_t`
+    primitive_kind_t primitive_kind_;
+
+    // Obtained from `engine_t`
+    engine_kind_t engine_kind_;
+    runtime_kind_t runtime_kind_;
+    device_id_t device_id_;
+
+    // Obtained from ctor
+    int impl_nthr_;
+
+    // ... //
+```
+
+When creating a key to find a primitive in the cache there is a primitive descriptor
+which is used to obtain some of the required data. However, in case of primitive descriptor
+creation for all primitives except reorder, concat and sum user creates `op_desc_t`,
+`primitive_attr_t`, `engine_t` and optionally a hint primitive descriptor. Therefore
+`impl_id_`, `bwd_mds` and `op_desc_t` for the aforementioned primitives cannot
+be obtained at that point.
+
+#### Implementation ID
+
+The rationale behind having an implementation ID is to be able to differentiate
+the same primitives which have different underlying implementations. Such primitives
+can be created when user uses primitive descriptor iterator to manually iterate
+through the list of implementations.
+
+The problem here is that the primitive descriptor is only going to be created and
+currently there is no way to get implementation ID without creating primitive
+descriptor.
+
+Currently, `std::type_index` of a primitive descriptor of a particular implementation
+is used as the implementation ID. In general, this approach has only one minor issue,
+it doesn't make it possible to build oneDNN without RTTI support, though it's
+unclear whether such ability is needed. However, to be able to use the existing
+key and hence the existing cache for primitives to speed up primitive descriptor
+creation the implementation ID must be reworked.
+
+The idea is to replace the existing implementation ID with a new one which is
+defined as follows:
+
+Implementation ID is a relative index of a valid implementation for the given
+operation descriptor, attributes and a forward primitive descriptor hint.
+
+For example there are operation descriptor and attributes for which some of
+the implementations in the implementation list are valid. Then the implementation
+IDs will be the following:
+
+| Implementation ID | Implementation index in the list | Valid |
+| ----------------- | -------------------------------- | ----- |
+|       -           |                0                 |   -   |
+|       0           |                1                 |   +   |
+|       -           |                2                 |   -   |
+|       1           |                3                 |   +   |
+|       -           |                4                 |   -   |
+|       2           |                5                 |   +   |
+
+In other words, the implementation IDs are the indexes in a hypothetical list
+of valid implementations for the given operation descriptors and attributes.
+
+The implementation ID can be non-zero only when using a primitive descriptor
```

Comment:
How do you think we should handle the situation below (let's assume an empty cache for simplicity):
~~~cpp
auto pd = convolution_forward::primitive_desc(op_desc, attr, eng); // ----> cache_miss, ID = 0
pd.next_impl(); // ----> cache miss, ID = 1
auto prim = convolution_forward(pd); // ---> cache_miss, ID = 1, cache entry created
auto pd = convolution_forward::primitive_desc(op_desc, attr, eng); // ----> cache_miss or cache_hit?
~~~

In other words, should we extract the primitive descriptor implementation from the cache regardless of whether the user has iterated using `next_impl()`?

If we only extract the implementation ID that the user-facing `primitive_desc` object contains, then the execution will always have to iterate over the implementation list until finding the one available in the primitive cache.

Reply:
In the code snippet it will a cache miss because it will start with the first valid implementation but it was the second implementation that was cached.

Although, iterating through the list of implementations will not benefit much from approach described in this RFC it's not the common case we are trying to optimize. Basically, frameworks and other users do not use the iterator unless they need a workaround to fall back to a reference/simple implementation. But in this case the performance will suffer anyway.